src/System.Private.CoreLib/src/System/Threading/Tasks/TaskFactory.cs

```diff
@@ -2291,11 +2291,13 @@ internal sealed class CompleteOnInvokePromise : Task<Task>, ITaskCompletionActio
         {
             private IList<Task> _tasks; // must track this for cleanup
             private int m_firstTaskAlreadyCompleted;
+            private bool _isSyncBlocking;
```

Comment:
This will end up increasing the size of the allocation employed by all Task.WhenAnys on 32-bit by 4 bytes.  Is it worth the savings for Task.WaitAny?

Reply:
What about a new type? e.g. represent it as sync blocking with a different type rather than a flag?

Reply:
I think it'd be fine to either do that (ideally just deriving from this one to share 99% of the code and just overriding InvokeMayRunArbitraryCode) or just use a bit from m_firstTaskAlreadyCompleted.  I lean towards the latter; it'll just involve some bit masking / ors / ands.